src/NuGet.Core/NuGet.Packaging/Signing/Archive/ReadOnlyBufferedStream.cs

```diff
@@ -0,0 +1,365 @@
+            }
+
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            if (count - offset > buffer.Length ||
+                count + offset > buffer.Length)
+            {
+                throw new ArgumentException(
+                    string.Format(
+                        CultureInfo.CurrentCulture,
+                        Strings.RangeOutOfBoundsForArray,
+                        nameof(offset),
+                        nameof(count)));
+            }
+
+            var bytesRead = 0;
+            var destinationOffset = offset;
+            var bytesWanted = count;
+
+            while (bytesWanted > 0 && !IsPositionAfterEndOfStream())
+            {
+                int sourceStart;
+                int bytesToCopy;
+
+                if (IsPositionInBuffer())
+                {
+                    sourceStart = (int)(Position - _bufferStartPosition);
+                }
+                else
+                {
+                    FillBuffer();
+
+                    sourceStart = 0;
+                }
+
+                bytesToCopy = (int)Math.Min(_bufferStartPosition + _bufferFillLength - Position, bytesWanted);
+
+                Debug.Assert(bytesToCopy >= 0);
+
+                if (bytesToCopy <= 0)
+                {
+                    break;
+                }
+
+                Buffer.BlockCopy(_buffer, sourceStart, buffer, destinationOffset, bytesToCopy);
+
+                destinationOffset += bytesToCopy;
+                bytesRead += bytesToCopy;
+                bytesWanted -= bytesToCopy;
+
+                Position += bytesToCopy;
+            }
+
+            return bytesRead;
+        }
+
+        public override int ReadByte()
+        {
+            ThrowIfDisposed();
+
+            int offset;
+
+            if (IsPositionInBuffer())
+            {
+                offset = (int)(Position - _bufferStartPosition);
+            }
+            else if (IsPositionAfterEndOfStream())
+            {
+                return -1;
+            }
+            else
+            {
+                FillBuffer();
+
+                offset = 0;
+            }
+
+            var value = _buffer[offset];
+
+            ++Position;
+
+            return value;
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            ThrowIfDisposed();
+
+            long newOffset;
+
+            switch (origin)
+            {
+                case SeekOrigin.Current:
+                    newOffset = Position + offset;
+                    break;
+
+                case SeekOrigin.Begin:
+                    newOffset = offset;
+                    break;
+
+                case SeekOrigin.End:
+                    newOffset = Length + offset;
+                    break;
+
+                default:
+                    throw new ArgumentException();
+            }
+
+            if (newOffset < 0)
```

Comment:
Should this be `newOffset < 0 || newOffset >= Length`?

Reply:
No, it's legal to seek beyond the length of a stream.